pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceReplicaPartitionSelector.java

```diff
@@ -0,0 +1,187 @@
+        int pool = pools.get((_tableNameHash + replicaId) % numPools);
+        poolToReplicaIdsMap.computeIfAbsent(pool, k -> new ArrayList<>()).add(replicaId);
+      }
+      LOGGER.info("Selecting {} replicas from pool: {} for table: {}", numReplicas, poolToReplicaIdsMap,
+          _tableNameWithType);
+
+      int numInstancesPerReplica = _replicaPartitionConfig.getNumInstancesPerReplica();
+      if (numInstancesPerReplica > 0) {
+        // Check if we have enough instances if number of instances per replica is configured
+        for (Map.Entry<Integer, List<Integer>> entry : poolToReplicaIdsMap.entrySet()) {
+          int pool = entry.getKey();
+          int numInstancesInPool = poolToInstanceConfigsMap.get(pool).size();
+          int numInstancesToSelect = numInstancesPerReplica * entry.getValue().size();
+          Preconditions.checkState(numInstancesToSelect <= numInstancesInPool,
+              "Not enough qualified instances from pool: %s (%s in the pool, asked for %s)", pool, numInstancesInPool,
+              numInstancesToSelect);
+        }
+      } else {
+        // Use as many instances as possible if number of instances per replica is not configured
+        numInstancesPerReplica = Integer.MAX_VALUE;
+        for (Map.Entry<Integer, List<Integer>> entry : poolToReplicaIdsMap.entrySet()) {
+          int pool = entry.getKey();
+          int numReplicasInPool = entry.getValue().size();
+          int numInstancesInPool = poolToInstanceConfigsMap.get(pool).size();
+          Preconditions.checkState(numReplicasInPool <= numInstancesInPool,
+              "Not enough qualified instances from pool: %s, cannot select %s replicas from %s instances", pool,
+              numReplicasInPool, numInstancesInPool);
+          numInstancesPerReplica = Math.min(numInstancesPerReplica, numInstancesInPool / numReplicasInPool);
+        }
+      }
+      LOGGER.info("Selecting {} instances per replica for table: {}", numInstancesPerReplica, _tableNameWithType);
+
+      String[][] replicaIdToInstancesMap = new String[numReplicas][numInstancesPerReplica];
+      for (Map.Entry<Integer, List<Integer>> entry : poolToReplicaIdsMap.entrySet()) {
+        List<InstanceConfig> instanceConfigsInPool = poolToInstanceConfigsMap.get(entry.getKey());
+        List<Integer> replicaIdsInPool = entry.getValue();
+
+        // Use round-robin to assign instances to each replica so that they get instances with similar picking priority
+        int instanceIdInPool = 0;
+        for (int instanceIdInReplica = 0; instanceIdInReplica < numInstancesPerReplica; instanceIdInReplica++) {
+          for (int replicaId : replicaIdsInPool) {
+            replicaIdToInstancesMap[replicaId][instanceIdInReplica] =
+                instanceConfigsInPool.get(instanceIdInPool++).getInstanceName();
+          }
+        }
+      }
+
+      // Assign instances within a replica to one partition if not configured
```

Comment:
I am a bit confused on the term that we are using.

`replica group - a set of instances that contains all segments for a partition or table`

`replica - we used to use this term for indicating replication factor for a segment (e.g. 3 replicas)`

Is this still correct in your comments? What do you mean by `instances within a replica`?

Reply:
The replica here refers to the table level replica. Added javadoc for the class